xdsl/irdl/declarative_assembly_format.py

```diff
@@ -118,20 +115,27 @@ def parse(
 
         # Infer operand types that should be inferred
         unresolved_operands = state.operands
-        assert isa(unresolved_operands, list[UnresolvedOperand])
+        assert isa(
+            unresolved_operands, list[UnresolvedOperand | list[UnresolvedOperand]]
+        )
         self.resolve_operand_types(state, op_def)
         operand_types = state.operand_types
-        assert isa(operand_types, list[Attribute])
+        assert isa(operand_types, list[Attribute | list[Attribute]])
 
         # Infer result types that should be inferred
         self.resolve_result_types(state, op_def)
         result_types = state.result_types
-        assert isa(result_types, list[Attribute])
+        assert isa(result_types, list[Attribute | list[Attribute]])
 
         # Resolve all operands
-        operands = parser.resolve_operands(
-            unresolved_operands, operand_types, parser.pos
-        )
+        operands: Sequence[SSAValue | Sequence[SSAValue]] = []
+        for uo, ot in zip(unresolved_operands, operand_types):
+            if isinstance(uo, list):
+                assert isinstance(ot, list), ot
```

Comment:
feels like these should be parsing errors but I'm not sure how important that is

Reply:
This holds by construction, it's pyright handshaking; it's one of those cases where having this being false is not an input case, but properly a bug of the preceding steps.
I feel like throwing errors for the sake of prettifying some code present for static analysis might be sending the wrong message to later code readers :thinking: 
I guess a silent assert is not saying much more though. WDYT, should it just be a cast in that case? Then catching the bug is confusing :smiling_face_with_tear:

Reply:
I think it's good, maybe worth adding this explanation in the string next to the assert, instead of `ot`?

Reply:
At this point it's more of a nit

Reply:
Oh right, that much is just uncleaned assert debugging ahah. Thanks for the catch!